django/db/backends/utils.py

```diff
@@ -135,8 +135,10 @@ def typecast_timestamp(s):  # does NOT store time zone information
     # "2005-07-29 09:56:00-05"
     if not s:
         return None
-    if ' ' not in s:
+    if ':' not in s:
         return typecast_date(s)
+    if '-' not in s:
+        return typecast_time(s)
```

Comment:
To be honest, I am not sure if this should be the expected behavior. Right now it seems like we are making it so that if the input contains only a date or time, then call the respective typecast function and return that result. However, this means we cannot guarantee the return type of the output of typecast_timestamp, which I think was causing the other issues before. 

Maybe it would be better to make it an error to call it on a string that contains only date or only time information, and possibly rewrite the sql-generation to account for this (and only call typecast_date/typecast_time in those cases).